single_kernel_mongo/core/k8s_worload.py

```diff
@@ -0,0 +1,168 @@
+            logger.exception(str(e))
+            raise WorkloadServiceError(e.err) from e
+
+    @override
+    def restart(self) -> None:
+        try:
+            self.container.add_layer(self.layer_name, self.layer, combine=True)
+            self.container.restart(self.service)
+        except ChangeError as e:
+            logger.exception(str(e))
+            raise WorkloadServiceError(e.err) from e
+
+    @override
+    def mkdir(self, path: Path, make_parents: bool = False) -> None:
+        self.container.make_dir(path, make_parents=make_parents)
+
+    @override
+    def exists(self, path: Path) -> bool:
+        return self.container.exists(path)
+
+    @override
+    def read(self, path: Path) -> list[str]:
+        if not self.container.exists(path):
+            return []
+        with self.container.pull(path) as f:
+            return f.read().split("\n")
+
+    @override
+    def write(self, path: Path, content: str, mode: str = "w") -> None:
+        self.container.push(
+            path,
+            content,
+            make_dirs=True,
+            permissions=0o400,
+            user=self.users.user,
+            group=self.users.group,
+        )
+
+    @override
+    def delete(self, path: Path):
+        self.container.remove_path(path)
+
+    @override
+    def copy_to_unit(self, src: Path, destination: Path):
+        license_file = self.container.pull(path=src)
+        destination.write_text(license_file.read())
+
+    @override
+    def get_env(self) -> dict[str, str]:
+        return (
+            self.container.get_plan()
+            .to_dict()
+            .get("services", {})
+            .get(self.service, {})
+            .get("environment", {})
+        )
+
+    @override
+    def update_env(self, parameters: chain[str]):
+        self._env = " ".join(parameters)
+
+    @override
+    def exec(
+        self,
+        command: list[str],  # type: ignore[override]
+        env: dict[str, str] | None = None,
+        working_dir: str | None = None,
+    ) -> str:
+        try:
+            process = self.container.exec(
+                command=command,
+                environment=env,
+                working_dir=working_dir,
+                combine_stderr=True,
+            )
+            output, _ = process.wait_output()
+            return output
+        except ExecError as e:
+            logger.debug(e)
+            raise WorkloadExecError(
+                e.command,
+                e.exit_code,
+                e.stdout,
+                e.stderr,
+            ) from e
+
+    @override
+    def run_bin_command(
+        self,
+        bin_keyword: str,
+        bin_args: list[str] | None = None,
+        environment: dict[str, str] | None = None,
+    ) -> str:
+        bin_args = bin_args or []
+        environment = environment or {}
+        command = [f"{self.paths.binaries_path}/{self.bin_cmd}", bin_keyword, *bin_args]
+        return self.exec(command=command, env=environment or None)
```

Comment:
Because this one call `self.exec` so it's already handled in the one above so there's no need to handle it twice, the correct, unified and catchable error is raised already before.
`run_bin_command` is a helper to run the command associated to a workload (eg: `charmed-mongodb.mongosh` for `MongoDBWorkload`)